### PR TITLE
Mon 16050 periodic log flush

### PR DESCRIPTION
**Pull request by jean-christophe81**
## Description

log_v2 (engine and broker) are flushed by an asio timer

**Fixes** # (issue)
sometimes, logs are not flushed

## Type of change

- [X] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 21.10.x
- [X] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x (master)

<h2> How this pull request can be tested ? </h2>

Please describe the **procedure** to verify that the goal of the PR is matched. Provide clear instructions so that it can be **correctly tested**.

Any **relevant details** of the configuration to perform the test should be added.

## Checklist

- [ ] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: broker/core/src/config/applier/endpoint.cc [hunk 1]
```diff
@@ -231,7 +231,9 @@ void endpoint::_discard() {
 
   // Stop multiplexing.
   try {
-    multiplexing::engine::instance().stop();
+    auto eng = multiplexing::engine::instance_ptr();
+    if (eng)
+      eng->stop();
```
Comment (bouda1) [inline broker/core/src/config/applier/endpoint.cc:236]: again, why is it mandatory to check eng?

**Review by bouda1 — CHANGES_REQUESTED** [review]

File: broker/core/src/pool.cc [hunk 2]
```diff
@@ -58,7 +60,7 @@ void pool::unload() {
  * @return the IO context.
  */
 asio::io_context& pool::io_context() {
-  return instance()._io_context;
+  return *instance()._io_context;
```
[the remark this answers is not in the document]
Comment (jean-christophe81) [reply broker/core/src/pool.cc:63]: the * concern the _io_context and not the return type of instance()

File: broker/core/src/config/applier/endpoint.cc (hunk 1 above)
Comment (jean-christophe81) [reply broker/core/src/config/applier/endpoint.cc]: removed

Comment (jean-christophe81) [reply broker/core/src/config/applier/endpoint.cc]: removed

File: broker/core/src/multiplexing/engine.cc [hunk 3]
```diff
@@ -290,64 +278,88 @@ std::string engine::_cache_file_path() const {
...
+    callback_caller(send_to_mux_callback_type&& callback,
+                    const std::shared_ptr<engine>& parent)
+        : _callback(callback), _parent(parent) {}
+    ~callback_caller() {
+      bool expected = true;
+      _parent->_sending_to_subscribers.compare_exchange_strong(expected, false);
+      _parent->_send_to_subscribers([]() {});
+      _callback();
+    }
+  };
```
[the remark this answers is not in the document]
Comment (jean-christophe81) [reply broker/core/src/multiplexing/engine.cc]: commented

File: broker/core/src/multiplexing/engine.cc [hunk 4]
```diff
@@ -290,64 +278,88 @@ std::string engine::_cache_file_path() const {
...
+
+  bool expected = false;
+  if (!_sending_to_subscribers.compare_exchange_strong(expected, true)) {
```
[the remark this answers is not in the document]
Comment (jean-christophe81) [reply broker/core/src/multiplexing/engine.cc:332]: comment added

File: broker/core/src/multiplexing/publisher.cc [hunk 5]
```diff
@@ -69,7 +72,10 @@ int32_t publisher::write(const std::shared_ptr<io::data>& d) {
  * @return The number of events published.
  */
 int publisher::write(const std::list<std::shared_ptr<io::data>>& to_publish) {
-  engine::instance().publish(to_publish);
+  auto eng = engine::instance_ptr();
+  if (eng) {
+    eng->publish(to_publish);
```
[the remark this answers is not in the document]
Comment (jean-christophe81) [reply broker/core/src/multiplexing/publisher.cc]: removed

**Review by bouda1 — APPROVED** [review]
There are one or two comments that could be fixed just before the merge.

File: engine/enginerpc/engine_impl.cc [hunk 6]
```diff
@@ -22,6 +22,8 @@
 #include <sys/types.h>
 #include <future>
 
+#include <asio.hpp>
+
```
[the remark this answers is not in the document]
Comment (jean-christophe81) [reply engine/enginerpc/engine_impl.cc:26]: yes as asio is now used by log_v2

File: broker/core/src/main.cc [hunk 7]
```diff
@@ -283,8 +289,10 @@ int main(int argc, char* argv[]) {
         }
         log_v2::core()->info("main: termination request received by process");
       }
+      log_v2::instance().stop_flush_timer();
       // Unload endpoints.
       config::applier::deinit();
+      spdlog::shutdown();
```
[the remark this answers is not in the document]
Comment (jean-christophe81) [reply broker/core/src/main.cc]: moved

**Pull request merged**